firestore/client/field_path.cc

```diff
@@ -0,0 +1,97 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "firestore/client/field_path.h"
+
+namespace firestore {
+std::regex FieldPath::simple_field_name("[_a-zA-Z][_a-zA-Z0-9]*");
+
+FieldPath::FieldPath(std::vector<std::string> const& parts) : parts_(parts) {
+  for (const auto part : parts) {
+    if (part.empty()) {
```

Comment:
Shouldn't this check for valid characters across the board?

Reply:
invalid characters are only checked in the from_string.  It is a specification detail from firestore.